File: legacy/scripts/scan/area_resolver.py
"""F-037 — feature.modules[] → AreaRecord cluster.

Pure deterministic mapping. Reuses ``structure`` (passed in) to avoid a
second walk of the project tree.
"""

from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from pathlib import Path


_CLUSTER_DEPTH = 2  # group modules sharing the first N path segments
_SOURCE_EXT_CANDIDATES = (".py", ".ts", ".tsx", ".js", ".jsx", ".rs", ".go", ".md")


@dataclass(frozen=True)
class AreaRecord:
    slug: str
    label: str
    paths: tuple[str, ...]
    modules: tuple[str, ...]
    feature_id: str


def resolve_areas(
    feature: dict,
    *,
    project_root: Path,
    structure: dict,
) -> list[AreaRecord]:
    """Return one AreaRecord per cluster of resolvable modules.

    Mapping rules (deterministic):
        1. dotted/slashed module exact match (file or dir).
        2. fall back to module + common source extension (``.py`` / ``.ts`` etc).
        3. bare-name module → prefix-match against ``structure.entity_candidate_files``
           or ``structure.top_dirs``.
        4. unresolvable → yields a single ``unmapped-*`` area (paths empty).
    """
    raw_modules: list[str] = list(feature.get("modules") or [])
    feature_id = str(feature.get("id") or "")
    if not raw_modules:
        return []

    project_root = Path(project_root)

    resolved: list[tuple[str, str]] = []  # (module, resolved_relative_path)
    unmapped: list[str] = []

    top_dirs = set(structure.get("top_dirs") or [])
    candidate_files = list(structure.get("entity_candidate_files") or [])

    for module in raw_modules:
        path = _try_resolve(module, project_root, top_dirs, candidate_files)
        if path:
            resolved.append((module, path))
        else:
            unmapped.append(module)

    clusters: dict[tuple[str, ...], list[tuple[str, str]]] = {}
    for module, path in resolved:
        key = _cluster_key(path)
        clusters.setdefault(key, []).append((module, path))

    areas: list[AreaRecord] = []
    for key in sorted(clusters):
        items = sorted(clusters[key])
        modules_tuple = tuple(m for m, _ in items)
        paths_tuple = tuple(_dedupe_preserve_order(p for _, p in items))
        label = "/".join(key)
        areas.append(
            AreaRecord(
                slug=_slugify(label),
                label=label,
                paths=paths_tuple,
                modules=modules_tuple,
                feature_id=feature_id,
            )
        )

    for module in sorted(unmapped):
        digest = hashlib.sha1(module.encode("utf-8")).hexdigest()[:8]
        areas.append(
            AreaRecord(
                slug=f"unmapped-{digest}",
                label=f"unmapped:{module}",
                paths=(),
                modules=(module,),
                feature_id=feature_id,
            )
        )

    return areas


def _try_resolve(
    module: str,
    project_root: Path,
    top_dirs: set[str],
    candidate_files: list[str],
) -> str | None:
    if not module:
        return None

    if "/" in module or "." in module and not module.startswith("."):
        candidate = (project_root / module)
        if candidate.is_file():
            return module.replace("\\", "/")
        if candidate.is_dir():
            return module.replace("\\", "/")
        for ext in _SOURCE_EXT_CANDIDATES:
            with_ext = project_root / f"{module}{ext}"
            if with_ext.is_file():
                return f"{module}{ext}".replace("\\", "/")
        if "/" in module:
            parent = module.rsplit("/", 1)[0]
            if (project_root / parent).is_dir():
                return parent.replace("\\", "/")

    bare = module.split("/")[-1]
    for entry in candidate_files:
        if Path(entry).stem == bare:
            return entry
    if bare in top_dirs:
        return bare

    return None


def _cluster_key(rel_path: str) -> tuple[str, ...]:
    segments = tuple(rel_path.split("/")[:_CLUSTER_DEPTH])
    return segments


def _slugify(label: str) -> str:
    slug = re.sub(r"[^a-zA-Z0-9]+", "-", label).strip("-").lower()
    return slug or "area"


def _dedupe_preserve_order(values):
    seen: set = set()
    out: list = []
    for value in values:
        if value not in seen:
            seen.add(value)
            out.append(value)
    return out

File: src/scan/areaResolver.ts
/**
 * feature.modules → AreaRecord cluster mapping (F-101 port of
 * `scripts/scan/area_resolver.py`, F-037).
 *
 * Pure deterministic mapping. Reuses a {@link StructureSummary}
 * passed in by the caller to avoid a second walk of the project
 * tree.
 *
 * @module scan/areaResolver
 */

import {createHash} from 'node:crypto';
import {statSync} from 'node:fs';
import {join} from 'node:path';

import type {StructureSummary} from './structure.js';

const CLUSTER_DEPTH = 2;
const SOURCE_EXT_CANDIDATES: ReadonlyArray<string> = ['.py', '.ts', '.tsx', '.js', '.jsx', '.rs', '.go', '.md'];

/** One detected area cluster. */
export interface AreaRecord {
  slug: string;
  label: string;
  paths: ReadonlyArray<string>;
  modules: ReadonlyArray<string>;
  feature_id: string;
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

/** Returns one AreaRecord per cluster of resolvable modules. */
export function resolveAreas(
  feature: Record<string, unknown>,
  projectRoot: string,
  structure: StructureSummary,
): AreaRecord[] {
  const rawModules: string[] = [];
  if (Array.isArray(feature['modules'])) {
    for (const m of feature['modules']) {
      if (typeof m === 'string') {
        rawModules.push(m);
      } else if (isPlainObject(m) && typeof m['name'] === 'string') {
        rawModules.push(m['name']);
      }
    }
  }
  const featureId = typeof feature['id'] === 'string' ? feature['id'] : '';
  if (rawModules.length === 0) {
    return [];
  }

  const resolved: Array<[string, string]> = [];
  const unmapped: string[] = [];

  const topDirs = new Set<string>(structure.top_dirs);
  const candidateFiles = [...structure.entity_candidate_files];

  for (const module of rawModules) {
    const path = tryResolve(module, projectRoot, topDirs, candidateFiles);
    if (path !== null) {
      resolved.push([module, path]);
    } else {
      unmapped.push(module);
    }
  }

  const clusters = new Map<string, Array<[string, string]>>();
  for (const [module, path] of resolved) {
    const key = clusterKey(path).join('/');
    if (!clusters.has(key)) {
      clusters.set(key, []);
    }
    clusters.get(key)!.push([module, path]);
  }

  const areas: AreaRecord[] = [];
  for (const key of [...clusters.keys()].sort()) {
    const items = [...clusters.get(key)!].sort((a, b) => {
      if (a[0] !== b[0]) {
        return a[0] < b[0] ? -1 : 1;
      }
      return a[1] < b[1] ? -1 : a[1] > b[1] ? 1 : 0;
    });
    const modules = items.map(([m]) => m);
    const paths = dedupePreserveOrder(items.map(([, p]) => p));
    areas.push({
      slug: slugify(key),
      label: key,
      paths,
      modules,
      feature_id: featureId,
    });
  }

  for (const module of [...unmapped].sort()) {
    const digest = createHash('sha1').update(module).digest('hex').slice(0, 8);
    areas.push({
      slug: `unmapped-${digest}`,
      label: `unmapped:${module}`,
      paths: [],
      modules: [module],
      feature_id: featureId,
    });
  }

  return areas;
}

function tryResolve(
  module: string,
  projectRoot: string,
  topDirs: ReadonlySet<string>,
  candidateFiles: ReadonlyArray<string>,
): string | null {
  if (!module) {
    return null;
  }

  if (module.includes('/') || (module.includes('.') && !module.startsWith('.'))) {
    const candidate = join(projectRoot, module);
    if (isFile(candidate)) {
      return module.replace(/\\/g, '/');
    }
    if (isDirectory(candidate)) {
      return module.replace(/\\/g, '/');
    }
    for (const ext of SOURCE_EXT_CANDIDATES) {
      const withExt = join(projectRoot, `${module}${ext}`);
      if (isFile(withExt)) {
        return `${module}${ext}`.replace(/\\/g, '/');
      }
    }
    if (module.includes('/')) {
      const parent = module.slice(0, module.lastIndexOf('/'));
      if (isDirectory(join(projectRoot, parent))) {
        return parent.replace(/\\/g, '/');
      }
    }
  }

  const bare = module.split('/').pop()!;
  for (const entry of candidateFiles) {
    const stem = entry.split('/').pop()!.replace(/\.[^.]+$/, '');
    if (stem === bare) {
      return entry;
    }
  }
  if (topDirs.has(bare)) {
    return bare;
  }
  return null;
}

function clusterKey(relPath: string): string[] {
  return relPath.split('/').slice(0, CLUSTER_DEPTH);
}

function slugify(label: string): string {
  const slug = label.replace(/[^a-zA-Z0-9]+/g, '-').replace(/^-+|-+$/g, '').toLowerCase();
  return slug.length > 0 ? slug : 'area';
}

function dedupePreserveOrder<T>(values: ReadonlyArray<T>): T[] {
  const seen = new Set<T>();
  const out: T[] = [];
  for (const value of values) {
    if (!seen.has(value)) {
      seen.add(value);
      out.push(value);
    }
  }
  return out;
}
